Invoke proper caching on CI

miguelraz opened the pull request:
The previous structure of this workflow (.yml file) did not invoke the `Swatinem/rust-cache` with the appropriate workspace per job.

Let's leverage that for even more speedups 🏁 

Should fix #195

jonathanpallant commented:
Why aren't we using the test_fns file any more?

miguelraz commented:
2 main reasons:
* Many of those functions just became 1 liners anyway, so we might as well have them inside the CI script. This is not the case for last one that zips up and deploys it, so I think it makes sense for that to still live in `build_fns.sh`. 
* There's now 3 places that you have to keep track of if you (potentially) want to add a new crate: `build_fns.sh`, `parallel-build.yml` and the local version `build.sh`. I think it would be good to consolidate where possible.

jonathanpallant commented:
The original reason for having CI call build.sh was so that it was trivial to do the same build that CI would. It feels like we lost that in getting the parallelism.

I'd prefer to see most of the commands in the build_fns.sh file, and then have both the GitHub action and the build.sh file use it. Each Github action step doing 'doing the thing' should be a one-liner, and then build.sh can just call all the one-liners sequentially.

miguelraz commented:
Thanks for the context, I agree that we can recoup some of that context by having `build.sh` call all the functions defined in `build_fns.sh`.

I've also
* Added a doc comment template for adding a new job
* modified the weekly canary job to run hourly for faster debugging to make sure the `build.sh` is running properly.

I've added the second change out of precaution because, ATM, I'm facing a terminal hang just after the last command has finished zipping successfully, so I'm kind of stuck here:

```
...
2025-04-13 22:22:01 [INFO] (mdbook::book): Book building has started
Warning: The mdbook-mermaid preprocessor was built against version 0.4.36 of mdbook, but we're being called from version 0.4.40
2025-04-13 22:22:01 [INFO] (mdbook::book): Running the html backend
~/work/rust-exercises
^C

```

jonathanpallant commented:
Looks OK to me. Just a couple of notes:

1. Can you squash all the commits to this branch? Maybe leave as separate the one that bumps the cron job to hourly, so we can just drop that commit later.
2. Can you change the loopback-fw and puzzle-fw artefacts to only contain the loopback-fw file or the puzzle-fw file, and not the whole `./target/thumbv7em-none-eabihf/release` folder? For some reason there's a *huge* dependency in there (<https://crates.io/crates/fixed> and it's making the artefact zip files 60 MiB in size. I don't think the compiled dependencies are that useful.

jonathanpallant commented:
I'd like to get this in before tomorrow's training.
